### PR TITLE
Trunk ui

### PR DESCRIPTION
**Pull request by kepta**
Ok so this is the dreaded ui submodule. 
I really want you guys to rigorously test and scrutinise this PR as much as possible.

cc: @bhousel @tmcw @tyrasd

### CODE REVIEW

File: modules/ui/core/intro.js [hunk 1]
```diff
@@ -1,3 +1,8 @@
+import { Icon } from '../../svg/index';
+import { Entity, Graph } from '../../core/index';
+import { Browse } from '../../modes/index';
+import { Intro } from '../intro/index';
```
Comment (kepta) [inline modules/ui/core/intro.js]: I couldn't think of a better name. So added an uppercase.

File: modules/ui/core/intro.js [hunk 2]
```diff
@@ -93,7 +98,7 @@ export function intro(context) {
         }
 
         var steps = ['navigation', 'point', 'area', 'line', 'startEditing'].map(function(step, i) {
-            var s = iD.ui.intro[step](context, reveal)
+            var s = Intro[step](context, reveal)
```
Comment (kepta) [inline modules/ui/core/intro.js]: This is where `Intro` is being used.

File: modules/ui/core/intro.js [hunk 3]
```diff
@@ -154,3 +159,36 @@ export function intro(context) {
     }
     return intro;
 }
+
+intro.pointBox = function(point, context) {
```
Comment (kepta) [inline modules/ui/core/intro.js]: These guys are hanging like this. 
Should I export them ?

File: modules/ui/index.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
+import { ui } from './ui';
```
Comment (kepta) [inline modules/ui/index.js]: This is the part where it gets interesting.

File: modules/ui/index.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+} from './core/index';
+export {ui};
+
+ui.preset = preset;
+Object.keys(presetObj).map(function(p) {
+    ui.preset[p] = presetObj[p];
+});
+
+// Need to do for test cases
+ui.intro = intro;
```
Comment (kepta) [inline modules/ui/index.js]: I had to add these keys to ui as `tests` still use it.
If we can create  separate build for tests and have a node.env variable to conditionally bundle this code. We can very much remove it for our production build.

File: modules/ui/core/save.js [hunk 6]
```diff
@@ -1,4 +1,6 @@
+import { Save as SaveMode } from '../../modes/index';
```
Comment (kepta) [inline modules/ui/core/save.js]: Name check again.

File: modules/ui/core/preset.js [hunk 7]
```diff
@@ -1,3 +1,9 @@
+import { Icon } from '../../svg/index';
+import { TagReference } from './tag_reference';
+import { Browse } from '../../modes/index';
+import { presetObj } from '../preset/index';
```
Comment (kepta) [inline modules/ui/core/preset.js]: Again, I have this naming problem. So this time I named it `presetObj`. Let me know if something better fits.

File: modules/ui/intro/index.js [hunk 8]
```diff
@@ -1,5 +1,13 @@
...
+import { area } from './area';
+import { line } from './line';
+import { navigation } from './navigation';
+import { point } from './point';
+import { startEditing } from './start_editing';
+
+export var Intro = {
+    area: area,
+    line: line,
+    navigation: navigation,
```
Comment (kepta) [inline modules/ui/intro/index.js]: This I had to do as `core/intro.js` expects an object and then does `Intro[steps]`. 

Let me know what you think of this approach.

File: modules/ui/preset/index.js [hunk 9]
```diff
@@ -1,13 +1,36 @@
...
+import { text, url, number, email, tel } from './input';
+import { localized } from './localized';
+import { lanes } from './lanes';
+import { maxspeed } from './maxspeed';
+import { radio } from './radio';
+import { restrictions } from './restrictions';
+import { textarea } from './textarea';
+import { wikipedia } from './wikipedia';
+
+export var presetObj = {
```
Comment (kepta) [inline modules/ui/preset/index.js]: over here again I had to export them as an object, because of the way `core/preset.js` uses them.

**Comment by bhousel** [comment]
I don't understand why there are naming problems.   Can you elaborate?

**Comment by kepta** [comment]
Let's take core/intro.js. 
There world be two variables named intro.  One internal and one from ui/intro/index.js

**Comment by kepta** [comment]
closed in favour of #3222

**Pull request closed without merging**